build/MSBuildToolset/Program.cs

```diff
@@ -1,8 +1,107 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
-namespace DoNotUse
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Text;
+using Microsoft.Build.Construction;
+
+namespace MSBuildToolset
 {
-    public class Program
+    public sealed class Program
     {
-        public static int Main(string[] args) => 1;
+        public static int Main(string[] args)
+        {
+            if (args.Length != 3)
+            {
+                Console.WriteLine("Usage: RoslynRestore <solutionFile> <nugetExePath> <dotnetPath>");
+                return -1;
+            }
+
+            var nugetExePath = Path.GetFullPath(args[1]);
+            var dotnetPath = Path.GetFullPath(args[2]);
+
+            var solutionFile = SolutionFile.Parse(args[0]);
+            var projectPaths = solutionFile.ProjectsInOrder
+                .Select(p => p.AbsolutePath)
+                .Select(p => p.Replace('\\', '/'))
+                .Where(p => !p.EndsWith("shproj") && p.EndsWith("proj"))
+                .ToArray();
+
+            var nugetExeDir = Path.GetDirectoryName(nugetExePath);
+            string msbuildPath = Path.Combine(AppContext.BaseDirectory, "MSBuild.exe");
+            string entryPointTargetPath = Path.Combine(nugetExeDir,
+                "build/Targets/GetProjectsReferencingProjectJsonFilesEntryPoint.targets");
+            string afterBuildsTargetPath = Path.Combine(nugetExeDir,
+                "build/Targets/GetProjectsReferencingProjectJsonFiles.targets");
+            string resultsFilePath = Path.ChangeExtension(Path.GetTempFileName(), "dg");
+
+            var argumentBuilder = new StringBuilder(
```

Comment:
Can we add a description of why we are doing this?  I understand but want to make sure it's obvious to future developers.